primitives/staking/src/lib.rs

```diff
@@ -54,67 +56,93 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
+	/// non-staking locks on the bonded pair's balance this may not be accurate.
+	pub total: T::Balance,
+	/// The total amount of the stash's balance that will be at stake in any forthcoming
+	/// rounds.
+	pub active: T::Balance,
+}
+
+/// A generic representation of a staking implementation.
+///
+/// This interface uses the terminology of NPoS, but it is aims to be generic enough to cover other
+/// implementations as well.
 pub trait StakingInterface {
 	/// Balance type used by the staking system.
-	type Balance;
+	type Balance: PartialEq;
 
 	/// AccountId type used by the staking system
 	type AccountId;
 
-	/// The minimum amount required to bond in order to be a nominator. This does not necessarily
-	/// mean the nomination will be counted in an election, but instead just enough to be stored as
-	/// a nominator. In other words, this is the minimum amount to register the intention to
-	/// nominate.
-	fn minimum_bond() -> Self::Balance;
+	/// The minimum amount required to bond in order to set nomination intentions. This does not
+	/// necessarily mean the nomination will be counted in an election, but instead just enough to
+	/// be stored as a nominator. In other words, this is the minimum amount to register the
+	/// intention to nominate.
+	fn minimum_nominator_bond() -> Self::Balance;
 
-	/// Number of eras that staked funds must remain bonded for.
+	/// The minimum amount required to bond in order to set validation intentions.
+	fn minimum_validator_bond() -> Self::Balance;
+
+	/// Return a stash account that is controlled by a `controller`.
 	///
-	/// # Note
+	/// ## Note
 	///
-	/// This must be strictly greater than the staking systems slash deffer duration.
+	/// The controller abstraction is not permanent and might go away. Avoid using this as much as
+	/// possible.
+	fn stash_by_ctrl(controller: &Self::AccountId) -> Result<Self::AccountId, DispatchError>;
+
+	/// Number of eras that staked funds must remain bonded for.
 	fn bonding_duration() -> EraIndex;
 
 	/// The current era index.
 	///
 	/// This should be the latest planned era that the staking system knows about.
 	fn current_era() -> EraIndex;
 
-	/// The amount of active stake that `stash` has in the staking system.
-	fn active_stake(stash: &Self::AccountId) -> Option<Self::Balance>;
+	/// Returns the stake of `who`.
+	fn stake(who: &Self::AccountId) -> Result<Stake<Self>, DispatchError>;
 
-	/// The total stake that `stash` has in the staking system. This includes the
-	/// [`Self::active_stake`], and any funds currently in the process of unbonding via
-	/// [`Self::unbond`].
-	///
-	/// # Note
-	///
-	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
-	/// non-staking locks on the bonded pair's balance this may not be accurate.
-	fn total_stake(stash: &Self::AccountId) -> Option<Self::Balance>;
+	/// TODO: Possibly return a result here too for consistency.
```

Comment:
@kianenigma I think returning a result here is kinda weird, but at the same time if there is some actual error and we just swallow it by returning a false - it might be a bad thing. WDYT?

Reply:
No strong opinion -- both are fine for me.